pkg/index/job/save/usecase/save.go

```diff
@@ -0,0 +1,213 @@
+	server        starter.Server
+	indexer       service.Indexer
+}
+
+// New returns Runner instance.
+func New(cfg *config.Data) (_ runner.Runner, err error) {
+	eg := errgroup.Get()
+
+	dOpts, err := cfg.Save.Discoverer.Client.Opts()
+	if err != nil {
+		return nil, err
+	}
+	// skipcq: CRT-D0001
+	dOpts = append(dOpts, grpc.WithErrGroup(eg))
+
+	acOpts, err := cfg.Save.Discoverer.AgentClientOptions.Opts()
+	if err != nil {
+		return nil, err
+	}
+	// skipcq: CRT-D0001
+	acOpts = append(acOpts, grpc.WithErrGroup(eg))
+
+	discoverer, err := discoverer.New(
+		discoverer.WithAutoConnect(true),
+		discoverer.WithName(cfg.Save.AgentName),
+		discoverer.WithNamespace(cfg.Save.AgentNamespace),
+		discoverer.WithPort(cfg.Save.AgentPort),
+		discoverer.WithServiceDNSARecord(cfg.Save.AgentDNS),
+		discoverer.WithDiscovererClient(grpc.New(dOpts...)),
+		discoverer.WithDiscoverDuration(cfg.Save.Discoverer.Duration),
+		discoverer.WithOptions(acOpts...),
+		discoverer.WithNodeName(cfg.Save.NodeName),
+		discoverer.WithOnDiscoverFunc(func(ctx context.Context, c discoverer.Client, addrs []string) error {
+			last := len(addrs) - 1
+			for i := 0; i < len(addrs)/2; i++ {
+				addrs[i], addrs[last-i] = addrs[last-i], addrs[i]
+			}
+			return nil
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	indexer, err := service.New(
+		service.WithDiscoverer(discoverer),
+		service.WithSaveIndexingConcurrency(cfg.Save.Concurrency),
+		service.WithTargetAddrs(cfg.Save.TargetAddrs...),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	srv, err := starter.New(
+		starter.WithConfig(cfg.Server),
+		starter.WithGRPC(func(cfg *iconfig.Server) []server.Option {
+			return []server.Option{
+				server.WithGRPCOption(
+					grpc.ChainUnaryInterceptor(recover.RecoverInterceptor()),
+					grpc.ChainStreamInterceptor(recover.RecoverStreamInterceptor()),
+				),
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var obs observability.Observability
+	if cfg.Observability.Enabled {
+		obs, err = observability.NewWithConfig(
+			cfg.Observability,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &run{
+		eg:            eg,
+		cfg:           cfg,
+		observability: obs,
+		server:        srv,
+		indexer:       indexer,
+	}, nil
+}
+
+// PreStart is a method called before execution of Start, and it invokes the PreStart method of observability.
+func (r *run) PreStart(ctx context.Context) error {
+	if r.observability != nil {
+		return r.observability.PreStart(ctx)
+	}
+	return nil
+}
+
+// Start is a method used to initiate an operation in the run, and it returns a channel for receiving errors
+// during the operation and an error representing any initialization errors.
+func (r *run) Start(ctx context.Context) (<-chan error, error) {
+	ech := make(chan error, 4)
+	var sech, oech <-chan error
+	if r.observability != nil {
+		oech = r.observability.Start(ctx)
+	}
+	sech = r.server.ListenAndServe(ctx)
+	ipech, err := r.indexer.PreStart(ctx)
```

Comment:
Calling `PreStart` in `Start` might be confusing, so something like `StartMonitoring` should be better?

Reply:
Thank you for your suggestion!
How about StartClient? 🤔  since this is actually start of the client rather than monitoring?

Reply:
Sounds good.